pkg/defaults/defaults.go

```diff
@@ -0,0 +1,157 @@
+package defaults
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/operator-framework/operator-marketplace/pkg/apis/operators/v1"
+	wrapper "github.com/operator-framework/operator-marketplace/pkg/client"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+var (
+	// Dir is the directory where the default OperatorSource definitions are
+	// placed on disk. It will be empty if defaults are not required.
+	Dir string
+
+	// defaultsTracker is used to keep an in-memory record of default
+	// OperatorSources. It is a map of OperatorSource name to the OperatorSource
+	// definition in the defaults directory.
+	defaultsTracker = make(map[string]v1.OperatorSource)
+)
+
+// Defaults is the interface that can be used to ensure default OperatorSources
+// are always present on the cluster.
+type Defaults interface {
+	EnsureAll(client wrapper.Client) error
+	Ensure(client wrapper.Client, opsrcName string) error
+}
+
+type defaults struct {
+}
+
+// New returns a the singleton defaults
+func New() Defaults {
+	return &defaults{}
+}
+
+// Ensure checks if the given OperatorSource source is one of the
+// defaults and if it is, it ensures it is present on the cluster.
+func (d *defaults) Ensure(client wrapper.Client, opsrcName string) error {
+	opsrc, present := defaultsTracker[opsrcName]
+	if !present {
+		return nil
+	}
+
+	err := processOpSrc(client, opsrc)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// EnsureAll processes all the default OperatorSources and ensures they are
+// present on the cluster.
+func (d *defaults) EnsureAll(client wrapper.Client) error {
+	allErrors := []error{}
+	for name := range defaultsTracker {
+		err := d.Ensure(client, name)
+		if err != nil {
+			allErrors = append(allErrors, fmt.Errorf("Error handling %s - %v", name, err))
+		}
+	}
+	return utilerrors.NewAggregate(allErrors)
+}
+
+// PopulateTracker populates the defaultsTracker on initialization
+func PopulateTracker() error {
+	// Default directory has not been specified
+	if Dir == "" {
+		return nil
```

Comment:
should this return an error (or even a panic)? it seems like if we're ever calling this without a default directory, that's a bug.

Reply:
The [default value](https://github.com/operator-framework/operator-marketplace/pull/226/commits/cac4f5c349a4d3628f028857e09f4e8307fa6134#diff-9e63f7b0791edcb4a53c944552b88970R57) of `Dir` is blank so it works. I tested this out without passing a default and it worked. I just could not figure out an easy way to add an e2e.